python/triqs/gf/tools.py

```diff
@@ -259,8 +259,94 @@ def fit_legendre(g_t, order=10):
-    
+
     g_l = Gf(mesh=lmesh, target_shape=g_t.target_shape)
     g_l.data[:] = c_l.reshape(g_l.data.shape) / scale
 
     return g_l
+
+
+def discretize_bath(delta_inp, Nb , bandwidth = 3, t_val = 0.0, tol=1e-5):
+    """
+    discretizes a given Delta_iw with Nb bath sites
+
+    CAUTION: This function works only for 1x1 / scalar valued Gf objects.
+    if necessary call for each orbital in input Delta_iw separately.
+
+    Parameters
+    ----------
+    g : GfImFreq
+        Non-interacting Green's function.
+    Nb : int
+        number of bath sites
+    bandwidth : approximative bandwidth for init of bath energies
+    t_val : float
+        to which value should the bath hoppings be initialized
+    tol: float
+        tolerance for scipy minimize
+    Returns
+    -------
+    t_opt : list of optimized bath hopping parameters
+    e_opt : list of optimized bath energies
+    delta_iw : BlockGf (of GfImFreq) or GfImFreq
+               Hybridization function.
+    """
+    from scipy.optimize import minimize
+
+    ####
+    # define minimizer for scipy
+    def minimizer(parameters):
+        # first half of parameters are hoppings
+        t = parameters[0:Nb]
+        # scond half are bath energies
+        e = parameters[Nb:]
+
+        # build discretized bath function as
+        # delta = sum_nb t_n / (iw - eps_n)
+        for w in delta_disc.mesh:
+            delta_disc[w] = np.sum(t**2 / ( w.value - e ) ) - delta_inp[w]
+
+        # calculate norm sum_iw_n | delta_disc - delta_inp |^2
+        norm = (delta_disc*delta_disc.conjugate()).density().real
+
+        return norm
+    ####
+
+    # some tests if input is okay
+    assert not isinstance(delta_inp, BlockGf), 'delta_inp needs to be a GfImFreq not BlockGf'
+    assert isinstance(delta_inp.mesh, MeshImFreq), 'input delta_iw should have a mesh MeshImFreq'
+    assert len(delta_inp.target_shape)==0, 'delta_inp should have rank 0, i.e. a scalar valued Green function'
+
+    # prepare discretized delta with same mesh
+    delta_disc = delta_inp.copy()
+    delta_disc.zero()
+
+    # initialize bath_hoppings by setting to the provided value
+    bath_hoppings = np.array( Nb*[t_val] )
+
+    # bath energies are initialized as linspace over the approximate bandwidth
+    bath_energies = np.linspace(0, bandwidth, Nb)
+    # parameters for scipy are a 1D array
+    parameters = np.array( (bath_hoppings, bath_energies) )
+
+    # run the minimizer
+    result = minimize(minimizer, parameters)
+
+    if not result.success:
+        raise RuntimeError('optimization has failed: scipy minimize signaled no success.')
+
+    # results, taking absolute values of hoppings
+    e_opt = result.x[Nb:]
+    t_opt = np.abs(result.x[0:Nb])
+
+    # sort by energy
+    order = np.argsort(e_opt)
+    e_opt = e_opt[order]
+    t_opt = t_opt[order]
+
+    # create optimized discretized delta on same mesh
+    for w in delta_disc.mesh:
+        delta_disc[w] = np.sum(t_opt**2/( w.value - e_opt ) )
```

Comment:
```suggestion
```
You can skip this step as the minimizer will already have set delta_disc accordingly

Reply:
Hm. I thought so too, but in my example it did not work. delta_disc looks quite different if I construct it explicitly again from the parameters provided by the function. Maybe I did something wrong?

![image](https://user-images.githubusercontent.com/21337117/107239161-95bd1400-69f6-11eb-8fb1-875dcfc8e2a9.png)